File: android/app/src/main/java/com/boostgamefaster/GameDetectorModule.java
package com.boostgamefaster;

import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeArray;
import com.facebook.react.bridge.WritableNativeMap;

import java.util.List;

public class GameDetectorModule extends ReactContextBaseJavaModule {
    private final ReactApplicationContext reactContext;

    public GameDetectorModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @Override
    public String getName() {
        return "GameDetector";
    }

    @ReactMethod
    public void getInstalledGames(Promise promise) {
        try {
            PackageManager pm = reactContext.getPackageManager();
            List<ApplicationInfo> packages = pm.getInstalledApplications(PackageManager.GET_META_DATA);
            WritableArray gameList = new WritableNativeArray();

            for (ApplicationInfo app : packages) {
                // Check if the app belongs to CATEGORY_GAME or has game-like characteristics
                if ((app.category == ApplicationInfo.CATEGORY_GAME) ||
                    (app.metaData != null && app.metaData.containsKey("com.google.android.play.games")) ||
                    app.packageName.contains("game")) {
                    WritableMap gameInfo = new WritableNativeMap();
                    gameInfo.putString("name", pm.getApplicationLabel(app).toString());
                    gameInfo.putString("packageName", app.packageName);
                    gameInfo.putString("status", "Ready");
                    gameInfo.putString("resolution", "1920x1080"); // Default value
                    gameInfo.putString("fpsCap", "60"); // Default value
                    gameList.pushMap(gameInfo);
                }
            }
            promise.resolve(gameList);
        } catch (Exception e) {
            promise.reject("ERROR", "Failed to get installed games: " + e.getMessage());
        }
    }

    @ReactMethod
    public void launchGame(String packageName, Promise promise) {
        try {
            Intent launchIntent = reactContext.getPackageManager().getLaunchIntentForPackage(packageName);
            if (launchIntent != null) {
                launchIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                reactContext.startActivity(launchIntent);
                promise.resolve("Launched " + packageName);
            } else {
                promise.reject("ERROR", "Could not launch game: " + packageName);
            }
        } catch (Exception e) {
            promise.reject("ERROR", "Failed to launch game: " + e.getMessage());
        }
    }
}

File: android/app/src/main/java/com/boostgamefaster/GameDetectorModule.kt
package com.boostgamefaster

import android.content.Context
import android.content.pm.PackageManager
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.Callback
import com.facebook.react.modules.core.DeviceEventManagerModule
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class GameDetectorModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    private val context: Context = reactContext
    private val coroutineScope = CoroutineScope(Dispatchers.Main)

    override fun getName(): String = "GameDetector"

    /**
     * Retrieves installed games with validation against known packages.
     * @param callback Returns list of games or error
     */
    @ReactMethod
    fun getInstalledGames(callback: Callback) {
        coroutineScope.launch {
            try {
                val packageManager = context.packageManager
                val installedApps = packageManager.getInstalledApplications(PackageManager.GET_META_DATA)
                val validGames = listOf(
                    "com.tencent.ig", // PUBG Mobile
                    "com.activision.callofduty.shooter", // Call of Duty Mobile
                    "com.miHoYo.GenshinImpact" // Genshin Impact
                )
                val games = installedApps
                    .filter { it.packageName in validGames || it.packageName.contains("game") || it.packageName.contains("play") }
                    .map {
                        mapOf(
                            "name" to sanitizeInput(packageManager.getApplicationLabel(it).toString()),
                            "packageName" to it.packageName,
                            "genre" to "Unknown" // Simplified, can be enhanced with Play Store API
                        )
                    }
                callback.invoke(null, games)
            } catch (e: Exception) {
                callback.invoke("Error: ${e.message}")
            }
        }
    }

    /**
     * Retrieves all installed apps for manual selection.
     * @param callback Returns list of apps or error
     */
    @ReactMethod
    fun getAllInstalledApps(callback: Callback) {
        coroutineScope.launch {
            try {
                val packageManager = context.packageManager
                val installedApps = packageManager.getInstalledApplications(PackageManager.GET_META_DATA)
                val apps = installedApps
                    .filter { it.packageName != context.packageName } // Exclude self
                    .map {
                        mapOf(
                            "name" to sanitizeInput(packageManager.getApplicationLabel(it).toString()),
                            "packageName" to it.packageName,
                            "genre" to "Unknown"
                        )
                    }
                callback.invoke(null, apps)
            } catch (e: Exception) {
                callback.invoke("Error: ${e.message}")
            }
        }
    }

    /**
     * Starts monitoring game status and emits updates.
     */
    @ReactMethod
    fun startMonitoring() {
        coroutineScope.launch {
            while (true) {
                val status = mapOf(
                    "cpuUsage" to (Math.random() * 100).toFloat(),
                    "ramUsage" to (Math.random() * 100).toFloat(),
                    "fps" to (Math.random() * 60).toInt(),
                    "ping" to (Math.random() * 100).toInt()
                )
                withContext(Dispatchers.Main) {
                    reactApplicationContext
                        .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                        .emit("GameStatusUpdate", status)
                }
                kotlinx.coroutines.delay(5000)
            }
        }
    }

    /**
     * Sanitizes input to prevent injection attacks.
     * @param input Input string
     * @return Sanitized string
     */
    private fun sanitizeInput(input: String): String {
        return input.replace(Regex("[<>{};]"), "")
    }
}
